libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/RouteAlternativesController.kt

```diff
@@ -0,0 +1,113 @@
+import com.mapbox.navigation.core.routeoptions.RouteOptionsUpdater
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.navigator.internal.MapboxNativeNavigator
+import com.mapbox.navigation.utils.internal.MapboxTimer
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArraySet
+
+internal class RouteAlternativesController(
+    private val options: RouteAlternativesOptions,
+    private val navigator: MapboxNativeNavigator,
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsUpdater: RouteOptionsUpdater,
+    private val logger: Logger
+) {
+    private val jobControl = ThreadController.getMainScopeAndRootJob()
+
+    private val mapboxTimer = MapboxTimer()
+    private val observers = CopyOnWriteArraySet<RouteAlternativesObserver>()
+    private var currentRequestId: Long? = null
+
+    fun register(routeAlternativesObserver: RouteAlternativesObserver) {
+        val needsToStartTimer = observers.isEmpty()
+        observers.add(routeAlternativesObserver)
+        if (needsToStartTimer) {
+            mapboxTimer.restartAfterMillis = options.intervalMillis
+            mapboxTimer.startTimer {
+                requestRouteAlternatives()
+            }
+        }
+    }
+
+    fun unregister(routeAlternativesObserver: RouteAlternativesObserver) {
+        observers.remove(routeAlternativesObserver)
+        if (observers.isEmpty()) {
+            stop()
+        }
+    }
+
+    fun unregisterAll() {
+        observers.clear()
+        stop()
+    }
+
+    private fun stop() {
+        mapboxTimer.stopJobs()
+        jobControl.job.cancelChildren()
+        currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+    }
+
+    private fun requestRouteAlternatives() {
+        if (directionsSession.routes.isEmpty()) {
+            return
+        }
+
+        val routeOptionsResult = routeOptionsUpdater.update(
+            directionsSession.getPrimaryRouteOptions(),
+            tripSession.getRouteProgress(),
+            tripSession.getEnhancedLocation()
+        )
+
+        when (routeOptionsResult) {
+            is RouteOptionsUpdater.RouteOptionsResult.Success -> {
+                currentRequestId?.let { directionsSession.cancelRouteRequest(it) }
+                currentRequestId = directionsSession.requestRoutes(
+                    routeOptionsResult.routeOptions,
+                    routesRequestCallback
+                )
+            }
+            is RouteOptionsUpdater.RouteOptionsResult.Error -> Unit
+        }
+    }
+
+    fun isDifferentRoute(directionsRoute: DirectionsRoute, callback: DifferentRouteCallback) {
+        jobControl.scope.launch {
+            val isDifferentRoute = navigator.isDifferentRoute(directionsRoute)
```

Comment:
If this is computationally expensive, let's use the worker thread, not the main thread.

Reply:
This is controlling the callback thread, which happens on the main thread. The thread running the computation runs on the navigator's thread

See the `MapboxNativeNavigatorImpl` `withContext(NavigatorDispatcher)`
``` kotlin
    override suspend fun isDifferentRoute(
        directionsRoute: DirectionsRoute
    ): Boolean = withContext(NavigatorDispatcher) {
```

Reply:
Ah, I missed this dispatcher :+1: